cmd/kubelet/app/server.go

```diff
@@ -377,116 +347,63 @@ func run(s *options.KubeletServer, kubeDeps *kubelet.KubeletDeps) (err error) {
-			if err == nil {
-				// Update s (KubeletServer) with new config from API server
-				s.KubeletConfiguration = *remoteKC
-				// Ensure that /configz is up to date with the new config
-				if cfgzErr != nil {
-					glog.Errorf("was unable to register configz before due to %s, will not be able to set now", cfgzErr)
-				} else {
-					setConfigz(cfgz, &s.KubeletConfiguration)
-				}
-				// Update feature gates from the new config
-				err = utilfeature.DefaultFeatureGate.Set(s.KubeletConfiguration.FeatureGates)
-				if err != nil {
-					return err
-				}
-			} else {
-				glog.Errorf("failed to init dynamic Kubelet configuration sync: %v", err)
-			}
-		}
-	}
-
-	// Validate configuration.
-	if err := validateConfig(s); err != nil {
-		return err
-	}
-
+	var nodeName types.NodeName
 	if kubeDeps == nil {
-		var kubeClient clientset.Interface
-		var eventClient v1core.EventsGetter
-		var externalKubeClient clientgoclientset.Interface
-		var cloud cloudprovider.Interface
-
-		if !cloudprovider.IsExternal(s.CloudProvider) && s.CloudProvider != componentconfigv1alpha1.AutoDetectCloudProvider {
-			cloud, err = cloudprovider.InitCloudProvider(s.CloudProvider, s.CloudConfigFile)
-			if err != nil {
-				return err
-			}
-			if cloud == nil {
-				glog.V(2).Infof("No cloud provider specified: %q from the config file: %q\n", s.CloudProvider, s.CloudConfigFile)
-			} else {
-				glog.V(2).Infof("Successfully initialized cloud provider: %q from the config file: %q\n", s.CloudProvider, s.CloudConfigFile)
-			}
+		cloud, err := getCloud(s)
+		if err != nil {
+			return err
 		}
-
-		nodeName, err := getNodeName(cloud, nodeutil.GetHostname(s.HostnameOverride))
+		nodeName, err = getNodeName(cloud, nodeutil.GetHostname(s.HostnameOverride))
 		if err != nil {
 			return err
 		}
-
 		if s.BootstrapKubeconfig != "" {
 			if err := bootstrapClientCert(s.KubeConfig.Value(), s.BootstrapKubeconfig, s.CertDirectory, nodeName); err != nil {
 				return err
 			}
 		}
-
-		clientConfig, err := CreateAPIServerClientConfig(s)
-		if err == nil {
-			kubeClient, err = clientset.NewForConfig(clientConfig)
-			if err != nil {
-				glog.Warningf("New kubeClient from clientConfig error: %v", err)
-			}
-			externalKubeClient, err = clientgoclientset.NewForConfig(clientConfig)
-			if err != nil {
-				glog.Warningf("New kubeClient from clientConfig error: %v", err)
-			}
-			// make a separate client for events
-			eventClientConfig := *clientConfig
-			eventClientConfig.QPS = float32(s.EventRecordQPS)
-			eventClientConfig.Burst = int(s.EventBurst)
-			eventClient, err = clientgoclientset.NewForConfig(&eventClientConfig)
-			if err != nil {
-				glog.Warningf("Failed to create API Server client: %v", err)
-			}
-		} else {
-			if s.RequireKubeConfig {
-				return fmt.Errorf("invalid kubeconfig: %v", err)
-			}
-			if standaloneMode {
-				glog.Warningf("No API client: %v", err)
-			}
+		kubeClient, _, _, err := getClients(s, standaloneMode)
```

Comment:
Distinct components that don't share variables or global state are way easier to reason about (even if they end up being run in sequence or in parallel in the same process), and separating them exposes data flow cycles and dependencies that we need to resolve, not just work around.

I also agree with @deads2k that air-gapping the config as a file the main kubelet run() method simply consumes helps inspectability and enables other config management integrations. I thought replacing flags with config files was the first step we were taking in order to get to dynamic config.

> You want clients to have the latest QPS, etc. settings. These settings are part of the configuration, and it makes reasonable sense to allow them to be dynamically updated.
> 
> You need to get a client to do an initial check for newer configuration when bootstrapping the controller. If, after the initial sync, you end up with a newer configuration than you started up with, you need to update the client to respect that. The simplest way to do this is to just regenerate the client. I guess we could try editing the parameters of the client in place, but this is less general, and maybe less safe.

It is the in-place mutation and side-effects that we should avoid. Rather than building a KubeletServer, then having it mutate itself and build and rebuild and overwrite clients, I'd expect startup to have a logical series of steps, each of which takes defined inputs and outputs, and does not mutate its inputs.

Ideally, each step would operate on the minimal set of inputs, but until we reach that factoring, I'd rather build all the inputs from scratch for each step and let each step run and produce the outputs it is responsible for:
1. client bootstrap produces kubeconfig
2. config bootstrap produces config file
3. server cert bootstrap produces serving cert
4. kubelet run consumes

No step should need to do anything differently based on how (or whether) the previous step ran